utils/dataset_manifest/utils.py

```diff
@@ -21,4 +24,161 @@ def rotate_image(image, angle):
+        name = os.path.splitext(os.path.basename(image_path))[0]
+        dirname = os.path.dirname(image_path)
+        related_images_dirname = os.path.normpath(os.path.join(dirname, '..', 'related_images'))
+        related_images[rel_image_path] = []
+
+        if latest_dirname != dirname:
+            # Update some data applicable for a subset of paths (within the current dirname)
+            latest_dirname = dirname
+            related_images_exist = os.path.isdir(related_images_dirname)
+            dirname_files = list(filter(lambda x: x != image_path, _list_and_join(dirname)))
+            velodyne_context_images_dirs = [directory for directory
+                in _list_and_join(os.path.normpath(os.path.join(dirname, '..', '..')))
+                if os.path.isdir(os.path.join(directory, 'data')) and re.search(r'image_\d.*', directory, re.IGNORECASE)
+            ]
+
+        if os.path.basename(dirname) == name:
+            # default format (option 2)
+            related_images[rel_image_path].extend(_prepare_context_list(dirname_files, root_path))
+
+        filtered_dirname_files = list(filter(lambda x: os.path.splitext(os.path.basename(x))[0] == name, dirname_files))
+        if len(filtered_dirname_files):
+            # default format (option 1)
+            related_images[rel_image_path].extend(_prepare_context_list(filtered_dirname_files, root_path))
+
+        if related_images_exist:
+            related_images_dirname = os.path.join(
+                related_images_dirname, '_'.join(os.path.basename(image_path).rsplit('.', 1))
+            )
+            if os.path.isdir(related_images_dirname):
+                related_images[rel_image_path].extend(
+                    _prepare_context_list(_list_and_join(related_images_dirname), root_path)
+                )
+
+        if dirname.endswith(os.path.join('velodyne_points', 'data')):
+            # velodynepoints format
+            for context_images_dir in velodyne_context_images_dirs:
+                context_files = _list_and_join(os.path.join(context_images_dir, 'data'))
+                context_files = list(
+                    filter(lambda x: os.path.splitext(os.path.basename(x))[0] == name, context_files)
+                )
+                related_images[rel_image_path].extend(
+                    _prepare_context_list(context_files, root_path)
+                )
+
+        related_images[rel_image_path].sort()
+    return related_images
+
+# This function is expected to be called only for images tasks
+# image_path is expected to be a list of absolute path to images
+# root_path is expected to be a string (dataset root)
+def detect_related_images(image_paths, root_path):
+    data_are_2d = False
+    data_are_3d = False
+
+    # First of all need to define data type we are working with
+    for image_path in image_paths:
+        # .bin files are expected to be converted to .pcd before this code
+        if os.path.splitext(image_path)[1].lower() == '.pcd':
```

Comment:
@bsekachev , from my perspective it is better to have an explicit flag. But let's live with the solution.